### PR TITLE
[sailfish-browser] User data initialization. Fixes JB#49922

### PR DESCRIPTION
**Pull request by mkenttala**

### CODE REVIEW

File: src/browser/browser.cpp [hunk 1]
```diff
@@ -22,13 +22,95 @@
...
     : view(view)
     , closeEventFilter(nullptr)
 {
+    initUserData();
+}
+
+void BrowserPrivate::initUserData()
+{
+    QDir cache(QDir::homePath() + MOZILLA_DATA_DIR + MOZILLA_DATA_STARTUPCACHE_DIR);
+    if (cache.exists())
```
Comment (rainemak) [inline src/browser/browser.cpp:46]: This was only run upon package upgrade not all the time.

File: src/browser/browser.cpp [hunk 2]
```diff
@@ -22,13 +22,90 @@
...
+            QFile source(MOZILLA_DATA_UA_UPDATE_SOURCE);
+            if (source.open(QIODevice::ReadOnly)) {
+                QByteArray line;
+                while (!source.atEnd()) {
+                    line = source.readLine();
+                    if (!line.trimmed().startsWith("//"))
+                        ua.write(line);
+                }
+                source.close();
+                ua.close();
```
Comment (okodron) [inline src/browser/browser.cpp:60]: I guess it should be in the upper scope where `ua.open()` is done.
However QFile object is created on stack and will be closed on destruction...

File: src/browser/browser.cpp [hunk 3]
```diff
@@ -22,13 +22,95 @@
...
+            }
+        } else {
+            qWarning() << "Could not open ua-update.json";
+        }
+    }
+
+    QFile prefs(QDir::homePath() + MOZILLA_DATA_DIR + MOZILLA_DATA_PREFS);
```
[the remark this answers is not in the document]
Comment (okodron) [reply src/browser/browser.cpp:67]: Can we play with the file modification timestamps? Like update prefs if the original source pref file is newer then the user's one.

[the remark this answers is not in the document]
Comment (rainemak) [reply src/browser/browser.cpp:67]: Or we even simpler hybrid between oneshot and copy.

File: rpm/sailfish-browser.spec [hunk 4]
```diff
@@ -117,11 +114,22 @@ install -m 644 -p %{SOURCE1} %{buildroot}%{_datadir}/mapplauncherd/privileges.d/
 
 # Upgrade, count is 2 or higher (depending on the number of versions installed)
 if [ "$1" -ge 2 ]; then
-%{_bindir}/add-oneshot --all-users --now browser-cleanup-startup-cache || :
```
Comment (rainemak) [inline rpm/sailfish-browser.spec:120]: Imo with approach it'd be cleaner to keep this one as a oneshot.

Comment (rainemak) [reply rpm/sailfish-browser.spec:120]: So let's keep "browser-cleanup-startup-cache" and schedule that for --all-users. If prefs and ua-updates do not exist those are/will be handled in the code.

File: rpm/sailfish-browser.spec [hunk 5]
```diff
@@ -117,11 +117,22 @@ install -m 644 -p %{SOURCE1} %{buildroot}%{_datadir}/mapplauncherd/privileges.d/
...
+    %{_bindir}/add-oneshot --all-users --now browser-cleanup-startup-cache || :
+    USERS=$(getent group users | cut -d ":" -f 4 | tr "," "\n")
```
Comment (rainemak) [inline rpm/sailfish-browser.spec:121]: Maybe run this as well with
```
%{_bindir}/add-oneshot --new-users --all-users --late browser-update-default-data || :
```

File: src/browser/browser.cpp [hunk 6]
```diff
@@ -22,13 +22,50 @@
...
+const auto MOZILLA_DATA_PREFS_SOURCE = QStringLiteral("/usr/share/sailfish-browser/data/prefs.js");
+
 BrowserPrivate::BrowserPrivate(QQuickView *view)
     : view(view)
     , closeEventFilter(nullptr)
 {
+    initUserData();
+}
+
+void BrowserPrivate::initUserData()
```
Comment (rainemak) [inline src/browser/browser.cpp:42]: This looks good now I think. It does exactly what it should i.e. if .mozembed is gone the data is copied back upon application startup.

**Comment by rainemak** [comment]
Please re-target this one to the next branch.

**Comment by rainemak** [comment]
Let's rebase sailfish-esr60 from next with --rebase-merges

**Comment by rainemak** [comment]
LGTM

**Review by rainemak — APPROVED** [review]
LGTM

**Pull request merged**